### PR TITLE
[MRG] Template alignment module and tests

### PR DESCRIPTION
**Pull request by thomasbazeille**

### CODE REVIEW

File: fmralign/alignment_methods.py [hunk 1]
```diff
@@ -135,7 +135,7 @@ def fit(self, X, Y):
     def transform(self, X):
         """Transform X using optimal transform computed during fit.
         """
-        return self.scale * X.dot(self.R)
+        return X.dot(self.R)
```
Comment (bthirion) [inline fmralign/alignment_methods.py]: Why do you remove the product by self.scale ?

File: fmralign/alignment_methods.py [hunk 2]
```diff
@@ -249,4 +249,4 @@ def fit(self, X, Y):
     def transform(self, X):
         """Transform X using optimal coupling computed during fit.
         """
-        return self.R.dot(X)
+        return X.dot(self.R)
```
Comment (bthirion) [inline fmralign/alignment_methods.py]: Do you have a specific test to check that these operations are done correctly ?

File: fmralign/examples/fetch_example_data.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
+import os
+from nilearn.datasets.utils import _fetch_files
+
+'''
+Example function from nilearn to download data from osf
```
Comment (bthirion) [inline fmralign/examples/fetch_example_data.py]: Put the description at the top of the modules

File: fmralign/examples/fetch_example_data.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+'''
+
+
+def fetch_localizer_button_task(data_dir=None, url=None, verbose=1):
```
Comment (bthirion) [inline fmralign/examples/fetch_example_data.py]: Note sure why you do this.
If this coy-pasted from nilearn or nistats, please add the dependency and do not introduce that function in this modules.
In any case, it should not be in the example folder, but part of the library.

File: fmralign/examples/fetch_example_data.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    dataset_name = 'brainomics'
+    data_dir = _get_dataset_dir(dataset_name, data_dir=data_dir,
+                                verbose=verbose)
+    files = _fetch_files(data_dir, filenames, verbose=verbose)
+
+    fdescr = _get_dataset_descr('brainomics_localizer')
+
+    params = dict([('description', fdescr)] + list(zip(options, files)))
+    return Bunch(**params)
```
Comment (bthirion) [inline fmralign/examples/fetch_example_data.py]: It's only one map per subject... I don't think it's relevant for hyperalignment, is it ?

File: fmralign/examples/fetch_example_data.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        verbosity level (0 means no message).
+    Returns
+    -------
+    data: Bunch
+        Dictionary-like object, the interest attributes are :
+        'tmap': string, giving paths to nifti contrast maps
+        'anat': string, giving paths to normalized anatomical image
+    Notes
+    ------
+    This function is only a caller for the fetch_localizer_contrasts in order
```
Comment (bthirion) [inline fmralign/examples/fetch_example_data.py]: to be removed

File: fmralign/examples/fetch_example_data.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    dataset_name = 'brainomics'
+    data_dir = _get_dataset_dir(dataset_name, data_dir=data_dir,
+                                verbose=verbose)
+    files = _fetch_files(data_dir, filenames, verbose=verbose)
+
+    fdescr = _get_dataset_descr('brainomics_localizer')
+
+    params = dict([('description', fdescr)] + list(zip(options, files)))
+    return Bunch(**params)
```
Comment (bthirion) [inline fmralign/examples/fetch_example_data.py]: Same comment heer. I'm missing the usecase for this function.

File: fmralign/examples/pairwise_alignment.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+from nilearn.datasets import fetch_neurovault, neurovault, fetch_neurovault_ids
```
Comment (bthirion) [inline fmralign/examples/pairwise_alignment.py]: OK, this looks unfinished. I think it is always good to start an example by describing your intention.

File: fmralign/pairwise_alignment.py [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+    Parameters
+    ----------
+    X: numpy array of shape (n_samples, n_features)
+        input data
+    mask: numpy array
+        the mask from which X is produced
+    n_pieces: int
+        number of different labels
+    clustering_method: string, optional
+        type of clustering applied to input data. Can be 'k_means', 'ward'
```
Comment (bthirion) [inline fmralign/pairwise_alignment.py]: defaults to k_means

File: fmralign/pairwise_alignment.py [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+    elif alignment_method == 'optimal_transport':
+        alignment_algo = OptimalTransportAlignment()
+    elif isinstance(alignment_method, (Identity, ScaledOrthogonalAlignment, RidgeAlignment, Hungarian, OptimalTransportAlignment)):
+        alignment_algo = copy.deepcopy(alignment_method)
+    alignment_algo.fit(X_i.T, Y_i.T)
+
+    return alignment_algo
+
+
+def fit_one_parcellation(X_, Y_, alignment_method, mask, n_pieces, clustering_method, joint_clustering, clustering_index, mem,
```
Comment (bthirion) [inline fmralign/pairwise_alignment.py]: In this function, please rename X_ and Y_ to X and Y

File: fmralign/pairwise_alignment.py [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+    ----------
+    X_: ndarray
+        Source data (shape : n_timeframes, n_voxels)
+    Y_: ndarray
+        Target data (shape : n_timeframes, n_voxels)
+    alignment_method: string
+        algorithm used to perform alignment between each region of X_ and Y_
+    mask: Niimg-like object
+        Mask to be used on data.
+    clustering_method : string
```
Comment (bthirion) [inline fmralign/pairwise_alignment.py]: n_pieces

File: fmralign/template_alignment.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+            X_norm += np.linalg.norm(img)
+        X_norm /= len(masked_imgs)
+        scale = X_norm / np.linalg.norm(average_img)
+    average_img *= scale
+
+    return masker.inverse_transform(mean_img)
+
+
+def reorder_img(img, labels, masker):
+    # img is 4D
```
Comment (bthirion) [inline fmralign/template_alignment.py]: short dosctring needed

File: fmralign/template_alignment.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+
+
+def align_template_to_images(template, imgs, method_alignment, n_pieces, n_bags, masker, n_jobs, reg):
+    '''
+    - Still should take care of clustering method/joint arguments
+    - Be sure that alignment_methods are the same in class docs and pairwise_alignment
+    - Additional arguments are available in pairwise_alignment : perturbation=False, smoothing_fwhm=None, standardize=None, detrend=False, target_affine=None, target_shape=None, low_pass=None, high_pass=None, t_r=None, memory=Memory(cachedir=None), memory_level=0.
+
+    same for align_images_to_template
+    '''
```
Comment (bthirion) [inline fmralign/template_alignment.py]: dosctring to be revised

File: fmralign/template_alignment.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+        template_copy = copy.copy(template)
+        piecewise_estimator = PairwiseAlignment(n_pieces=n_pieces, alignment_method=method_alignment,
+                                                mask=masker, clustering_method='k_means', joint_clustering=False, n_bags=n_bags, n_jobs=n_jobs)
+        piecewise_estimator.fit(template_copy, img)
+        deformation_operators.append(piecewise_estimator)
+        # data_csr = sparse.csr_matrix(data)
+    return deformation_operators
+
+
+def improve_template(deformation_operators, imgs, n_bags, scale_template, masker, n_jobs):
```
Comment (bthirion) [inline fmralign/template_alignment.py]: maybe 'template_update'
+  docstring

File: fmralign/template_alignment.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+    from scipy import sparse
+    regressed_templates, X_list = [], []
+    X_norm = 0
+    for img in imgs:
+        X_ = masker.transform(img)
+        X_list.append(X_)
+        X_norm += np.linalg.norm(X_)
+    X_stacked = np.hstack(X_list).T
+    X_norm /= len(X_list)
+    for bag in range(n_bags):
```
Comment (bthirion) [inline fmralign/template_alignment.py]: not sure what a bug reprsents in this function

File: fmralign/template_alignment.py [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+    if scale_template:
+        scale = X_norm / np.linalg.norm(averaged_regression)
+    return masker.inverse_transform(scale * averaged_regression)
+
+
+def create_template_regression(imgs, method_alignment, n_pieces, n_bags, scale_template, masker, n_iter, n_jobs, reg):
+    '''create_template deformed to match images s.t. :, at each iteration :
+        compute : T minimizing sum(||RT-X||) with least squares
+        compute new alignement transform images through transform R minimizing ||RT-X|| for each img
+    '''
```
Comment (bthirion) [inline fmralign/template_alignment.py]: same comment on doctring

File: fmralign/template_alignment.py [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+
+
+def align_images_to_template(imgs, template, method_alignment, n_pieces, n_bags, masker, n_iter, n_jobs, reg):
+    '''
+    - Still should take care of clustering method/joint arguments
+    - Be sure that alignment_methods are the same in class docs and pairwise_alignment
+    - Additional arguments are available in pairwise_alignment : perturbation=False, smoothing_fwhm=None, standardize=None, detrend=False, target_affine=None, target_shape=None, low_pass=None, high_pass=None, t_r=None, memory=Memory(cachedir=None), memory_level=0.
+
+    same for align_template_to_images
+    '''
```
Comment (bthirion) [inline fmralign/template_alignment.py]: again

File: fmralign/template_alignment.py [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+    aligned_imgs = []
+    for img in imgs:
+        piecewise_estimator = PairwiseAlignment(n_pieces=n_pieces, alignment_method=method_alignment,
+                                                mask=masker, clustering_method='k_means', joint_clustering=False, n_bags=n_bags, n_jobs=n_jobs)
```
Comment (bthirion) [inline fmralign/template_alignment.py]: line too long

File: fmralign/template_alignment.py [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+        piecewise_estimator.fit(img, template)
+        aligned_imgs.append(piecewise_estimator.transform(img))
+    return aligned_imgs
+
+
+def create_template(imgs, method_alignment, n_pieces, n_bags, scale_template, masker, n_iter, n_jobs, reg):
+    '''create_template by deforming images s.t. :, at each iteration :
+        compute : T minimizing sum(||RX-T||) which is the mean(RX)
+        compute new aligned images through transform R minimizing ||RX-T|| for each img
+    '''
```
Comment (bthirion) [inline fmralign/template_alignment.py]: again

File: fmralign/template_alignment.py [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+
+    mapping_image = index_img(img, train_index)
+    mapping = PairwiseAlignment(n_pieces=n_pieces, alignment_method=method_alignment, mask=masker,
+                                clustering_method='k_means', joint_clustering=False, n_bags=n_bags, n_jobs=n_jobs)
+    mapping.fit(mapping_image, img)
+    return mapping
+
+
+def predict_from_template_and_mapping(template, test_index,  mapping):
```
Comment (bthirion) [inline fmralign/template_alignment.py]: unclear name

File: fmralign/template_alignment.py [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+    image_to_transform = index_img(template, test_index)
+    transformed_image = mapping.transform(image_to_transform)
+    return transformed_image
+
+
+def compute_greedy_template(imgs, method_alignment, n_pieces, n_bags, masker, n_jobs):
```
Comment (bthirion) [inline fmralign/template_alignment.py]: 'geedy_template_estimation'

File: fmralign/template_alignment.py [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+    """
+    Greedy algorithm from Haxby 2011 to calculate a functional template from list of arrays X
+    implementation inspired of pymvpa
+
+    Parameters
+    ----------
+    X: list of array of shape n_features, n_timeframes
+    """
```
Comment (bthirion) [inline fmralign/template_alignment.py]: docstring to be done

File: fmralign/template_alignment.py [hunk 23]
```diff
@@ -0,0 +1,295 @@
...
+        realigned_imgs.append(piecewise_estimator.transform(imgs[i]))
+
+    return euclidian_mean_with_masking(realigned_imgs, False, masker)
+
+
+class TemplateAlignment(BaseEstimator, TransformerMixin):
+    """
+    Decompose the source and target images into source and target regions
+     Use alignment algorithms to align source and target regions independantly.
+    """
```
Comment (bthirion) [inline fmralign/template_alignment.py]: docstring

File: fmralign/tests/test_alignment_methods.py [hunk 24]
```diff
@@ -119,25 +119,31 @@ def test_optimal_permutation_on_translation_case():
...
-        assert_greater(algo_score, identity_baseline_score)
-
-    n_samples, n_features = 20, 100
-    X = np.random.randn(n_samples, n_features)
-    Y = np.random.randn(n_samples, n_features)
-
-    for algo in [ScaledOrthogonalAlignment(), RidgeAlignment(), OptimalTransportAlignment(), Hungarian()]:
-        assert_class_align_better_than_identity(algo, X, Y)
+        for algo in [RidgeAlignment(), ScaledOrthogonalAlignment(), OptimalTransportAlignment(), Hungarian()]:
+            print(algo)
```
Comment (bthirion) [inline fmralign/tests/test_alignment_methods.py]: likely to be removed

File: fmralign/alignment_methods.py (hunk 1 above)
Comment (thomasbazeille) [reply fmralign/alignment_methods.py]: Because I changed R to scale * R so that it represents the complete transform when queried through algo.R. It gives a behaviour more similar to other classes.

File: fmralign/alignment_methods.py (hunk 2 above)
Comment (thomasbazeille) [reply fmralign/alignment_methods.py]: Yes after finding a bug in the sense of the multiplication, I added a test of prediction shape to test_alignment_method to be sure that the product was done in the right way.

**Comment by thomasbazeille** [comment]
Good for review next week. Limited span PR but that is done properly I think. We can discuss enhancements or other usecases/examples but the basics are here I think.

**Comment by thomasbazeille** [comment]
This PR is ready for reviewing @bthirion

File: examples/template_alignment.py [hunk 25]
```diff
@@ -0,0 +1,164 @@
+# -*- coding: utf-8 -*-
+"""Creating a template for a group of subject using alignment
+===================================================================
+
+In this tutorial, we show how to better predict new contrasts for a target
+subject using many source subjects corresponding contrasts. For this purpose,
+we create and template to which we align the target subject.
```
Comment (bthirion) [inline examples/template_alignment.py]: a template

File: examples/template_alignment.py [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+
+We mostly rely on python common packages and on nilearn to handle
```
Comment (bthirion) [inline examples/template_alignment.py]: Python

File: examples/template_alignment.py [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+
+.. contents:: **Contents**
+    :local:
+    :depth: 1
+
+"""
+
+# Retrieving the data
+# -------------------
+# In this example we use the IBC dataset, which include a large number of \
```
Comment (bthirion) [inline examples/template_alignment.py]: includes

File: examples/template_alignment.py [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+# different contrasts maps for 12 subjects.
+# We download the images for subjects 1 2, 4, 5, 6 and 7 (or retrieve them \
+# if they were already downloaded).
+# Files is the list of paths for each subjects.
+# df is a dataframe with metadata about each of them.
+# mask is an appropriate nifti image to select the data.
```
Comment (bthirion) [inline examples/template_alignment.py]: mask is a binary image use to extract grey matter regions.

File: examples/template_alignment.py [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+#
+
+from fmralign.fetch_example_data import fetch_ibc_subjects_contrasts
+files, df, mask = fetch_ibc_subjects_contrasts(
```
Comment (bthirion) [inline examples/template_alignment.py]: should rename `mask` to `mask_img`

File: examples/template_alignment.py [hunk 30]
```diff
@@ -0,0 +1,164 @@
...
+# Defining a masker
+# -----------------
+# Using the mask provided, define a nilearn masker that will be used to \
+# handle relevant data. For more information, visit :
+# http://nilearn.github.io/manipulating_images/masker_objects.html
+#
+
+from nilearn.input_data import NiftiMasker
+masker = NiftiMasker(mask_img=mask)
+mask
```
Comment (bthirion) [inline examples/template_alignment.py]: ?

File: examples/template_alignment.py [hunk 31]
```diff
@@ -0,0 +1,164 @@
...
+masker.fit()
```
Comment (bthirion) [inline examples/template_alignment.py]: on one line:
```
 masker = NiftiMasker(mask_img=mask).fit()
```

File: examples/template_alignment.py [hunk 32]
```diff
@@ -0,0 +1,164 @@
...
+
+###############################################################################
+# Split the data
+# ---------------------------------------------
+# For each subject, we will use two 'copies' of data acquired in the same \
```
Comment (bthirion) [inline examples/template_alignment.py]: I don't like this wording: this gives the imrpession that hyperalignment only works if you have 2 copies of the data.
Just say two different series of contrast (we never explicitly use the fact that there is redundancy in the contrast specification).

File: examples/template_alignment.py [hunk 33]
```diff
@@ -0,0 +1,164 @@
...
+target_train = concat_imgs(target_train)
+target_test = df[df.subject == 'sub-07'][df.acquisition == 'pa'].path.values
+
+#############################################################################
+# Find the average of training subjects images, to be used as a baseline.
+# -------------------------------------------------------------------------
+# We use a function that return an image with as many contrasts as any subject
+# representing the average of all train subjects.
+
+from fmralign.template_alignment import euclidian_mean
```
Comment (bthirion) [inline examples/template_alignment.py]: should be euclidean I think

File: examples/template_alignment.py [hunk 34]
```diff
@@ -0,0 +1,164 @@
...
+template_train = []
+for i in range(5):
+    template_train.append(concat_imgs(files[i]))
+target_train = df[df.subject == 'sub-07'][df.acquisition == 'ap'].path.values
+# We make a single 4D Niimg from our list of 3D filenames
+target_train = concat_imgs(target_train)
+target_test = df[df.subject == 'sub-07'][df.acquisition == 'pa'].path.values
+
+#############################################################################
+# Find the average of training subjects images, to be used as a baseline.
```
Comment (bthirion) [inline examples/template_alignment.py]: Compute the average

File: examples/template_alignment.py [hunk 35]
```diff
@@ -0,0 +1,164 @@
...
+
+###############################################################################
+# Split the data
+# ---------------------------------------------
+# For each subject, we will use two 'copies' of data acquired in the same \
+# conditions during two independent sessions with a single varying parameter \
+# of acquisition, the phase encoding being either Antero-posterior(AP) or \
+# Postero-anterior(PA).
+#
+# We infer a template for subjects 1 to 6 for both AP and PA data.
```
Comment (bthirion) [inline examples/template_alignment.py]: sub-01 to sub-06

File: examples/template_alignment.py [hunk 36]
```diff
@@ -0,0 +1,164 @@
...
+#
+# For subject 7, we split it in two folds, one to learn alignement, the \
```
Comment (bthirion) [inline examples/template_alignment.py]: subject sub-07

File: examples/template_alignment.py (hunk 29 above)
Comment (bthirion) [reply examples/template_alignment.py]: should rename files to imgs

File: examples/template_alignment.py [hunk 37]
```diff
@@ -0,0 +1,164 @@
...
+#   the template
+# * target test: PA contrasts for subject 7, used as a ground truth to score \
+#   our predictions
+#
+from nilearn.image import concat_imgs
+
+# We make a list of 4D niimgs from our list of list of files containing 3D images
+template_train = []
+for i in range(5):
+    template_train.append(concat_imgs(files[i]))
```
Comment (bthirion) [inline examples/template_alignment.py]: you really need to learn concat_imgs ?

File: examples/template_alignment.py [hunk 38]
```diff
@@ -0,0 +1,164 @@
...
+    average_score), display_mode="z", vmax=1, cut_coords=[-15, -5])
+baseline_display.title("R2 score of prediction from group average")
+
+display = plotting.plot_stat_map(
+    masker.inverse_transform(
+        template_score), display_mode="z", cut_coords=[-15, -5], vmax=1)
+display.title("R2 score of prediction using template and alignment")
+plotting.show()
+#############################################################################
+# We observe that creating a template and aligning a new subject to it enables
```
Comment (bthirion) [inline examples/template_alignment.py]: aligning... to it yields  better prediction of his  test contrasts than just using the group average.

File: fmralign/template_alignment.py [hunk 39]
```diff
@@ -0,0 +1,393 @@
...
+
+
+def euclidian_mean(imgs, masker, scale_template=False):
+    """ Make the euclidian average of images.
+
+    Parameters
+    ----------
+    imgs: list of Niimgs
+        Each img is 3D by default, but can also be 4D.
+    masker:
```
Comment (bthirion) [inline fmralign/template_alignment.py]: docstring

File: fmralign/template_alignment.py [hunk 40]
```diff
@@ -0,0 +1,393 @@
...
+        piecewise_estimator.fit(img, template)
+        aligned_imgs.append(piecewise_estimator.transform(img))
+    return aligned_imgs
+
+
+def create_template(imgs, n_iter, scale_template, alignment_method, n_pieces,
+                    clustering_method, n_bags, masker, memory, memory_level,
+                    n_jobs, verbose):
+    '''Create template through alternate minimization.  Compute iteratively :
+        * T minimizing sum(||R_iX_i-T||) which is the mean of aligned images (RX_i)
```
Comment (bthirion) [inline fmralign/template_alignment.py]: (R_i X_i)

File: fmralign/template_alignment.py [hunk 41]
```diff
@@ -0,0 +1,393 @@
...
+        * align initial images to new template T
+            (find transform R minimizing ||R_iX_i-T|| for each img X_i)
```
Comment (bthirion) [inline fmralign/template_alignment.py]: transform R_i

File: fmralign/template_alignment.py [hunk 42]
```diff
@@ -0,0 +1,393 @@
...
+        alignment_method: string
+            Algorithm used to perform alignment between X_i and Y_i :
+            * either 'identity', 'scaled_orthogonal', 'ridge_cv',
+                'permutation', 'diagonal'
+            * or an instance of one of alignment classes
+                (imported from functional_alignment.alignment_methods)
+        n_pieces: int, optional (default = 1)
+            Number of regions in which the data is parcellated for alignment
+            If 1 the alignment is done on full scale data.
+            If >1, the voxels are clustered and alignment is performed
```
Comment (bthirion) [inline fmralign/template_alignment.py]: If > 1,

File: fmralign/template_alignment.py [hunk 43]
```diff
@@ -0,0 +1,393 @@
...
+
+        Parameters
+        ----------
+        alignment_method: string
+            Algorithm used to perform alignment between X_i and Y_i :
+            * either 'identity', 'scaled_orthogonal', 'ridge_cv',
+                'permutation', 'diagonal'
+            * or an instance of one of alignment classes
+                (imported from functional_alignment.alignment_methods)
+        n_pieces: int, optional (default = 1)
```
Comment (bthirion) [inline fmralign/template_alignment.py]: For this an other arguments, the defualt parameters are not specified in the function signature.

File: fmralign/template_alignment.py [hunk 44]
```diff
@@ -0,0 +1,393 @@
...
+        iter += 1
+
+    return template, template_history
+
+
+def map_template_to_image(img, train_index, template, alignment_method,
+                          n_pieces, clustering_method, n_bags, masker,
+                          memory, memory_level, n_jobs, verbose):
+    '''From a template, and new images, learn their alignment mapping
+    !!! mapping.fit seem to have wrong argument : if error it's upstream in functional_alignment.template
```
Comment (bthirion) [inline fmralign/template_alignment.py]: ?

File: fmralign/template_alignment.py [hunk 45]
```diff
@@ -0,0 +1,393 @@
...
+    - Be sure that alignment_methods are the same in class docs and pairwise_alignment
+    # Check everywhere whether template is list of 3D Niimgs or 4D Niimgs.
+
+
+    Parameters
+    ----------
+    img: list of 3D Niimgs
+
```
Comment (bthirion) [inline fmralign/template_alignment.py]: finish the dosctring

File: fmralign/template_alignment.py [hunk 46]
```diff
@@ -0,0 +1,393 @@
...
+        n_iter: int
+           number of iteration in the alternate minimization. Each img is
+           aligned n_iter times to the evolving template. If n_iter = 0,
+           the template is simply the mean of the input images.
+        save_template: None or string(optional)
+            If not None, path to which the template will be saved.
+
+        Returns
+        -------
+            self
```
Comment (bthirion) [inline fmralign/template_alignment.py]: wrong indent

File: fmralign/template_alignment.py [hunk 47]
```diff
@@ -0,0 +1,393 @@
...
+        TODO : test save_template
```
Comment (bthirion) [inline fmralign/template_alignment.py]: ?

File: fmralign/template_alignment.py [hunk 48]
```diff
@@ -0,0 +1,393 @@
...
+                                alignment_method=alignment_method,
+                                clustering_method=clustering_method,
+                                n_bags=n_bags, mask=masker, memory=memory,
+                                memory_level=memory_level,
+                                n_jobs=n_jobs, verbose=verbose)
+    mapping.fit(mapping_image, img)
+    return mapping
+
+
+def predict_from_template_and_mapping(template, test_index,  mapping):
```
Comment (bthirion) [inline fmralign/template_alignment.py]: All these functions should be private I think ?

File: fmralign/template_alignment.py [hunk 49]
```diff
@@ -0,0 +1,393 @@
...
+        Parameters
+        ----------
+        imgs: List of Niimg-like objects
+           See http://nilearn.github.io/manipulating_images/input_output.html
+           source data. Every img must have the same length (number of sample)
+        scale_template: boolean, default False
+            rescale template after each inference so that it keeps
+            the same norm as the average of training images.
+        n_iter: int
+           number of iteration in the alternate minimization. Each img is
```
Comment (bthirion) [inline fmralign/template_alignment.py]: iterations

File: fmralign/tests/test_pairwise_alignment.py [hunk 50]
```diff
@@ -22,9 +24,11 @@ def test_pairwise_identity():
                      'n_bags': 10, 'mask': mask_img, 'n_jobs': 2}
                  ]
     for args in args_list:
+        start = time.time()
         algo = PairwiseAlignment(**args)
         assert_algo_transform_almost_exactly(
             algo, img1, img1, mask=mask_img)
+        print(time.time() - start)
```
Comment (bthirion) [inline fmralign/tests/test_pairwise_alignment.py]: Why did you add that ?

File: examples/template_alignment.py (hunk 37 above)
Comment (thomasbazeille) [reply examples/template_alignment.py]: Yes for now to work with 4D images but will be modified by a PR answering issue #19

**Comment by thomasbazeille** [comment]
All comments of the previous review are taken into account @bthirion

File: fmralign/template_alignment.py [hunk 51]
```diff
@@ -0,0 +1,349 @@
+from sklearn.base import BaseEstimator, TransformerMixin
```
Comment (bthirion) [inline fmralign/template_alignment.py]: We need a module-level docstring

File: fmralign/template_alignment.py [hunk 52]
```diff
@@ -0,0 +1,349 @@
...
+import numpy as np
+from joblib import Parallel, delayed
+from sklearn.externals.joblib import Memory
+from nilearn.image import index_img
+from nilearn.input_data.masker_validation import check_embedded_nifti_masker
+from fmralign.pairwise_alignment import PairwiseAlignment
+
+
+def euclidean_mean(imgs, masker, scale_template=False):
```
Comment (bthirion) [inline fmralign/template_alignment.py]: if scale_template is False, then this is nothing else than Nielarn's `mean_img` I am a bit afraid of using `scale_template=True`, since the behavior becomes really different from 'euclidean mean'. 
At least , I'd rather call the function rescaled_mean to make its behavior clearer to users.

Comment (bthirion) [reply fmralign/template_alignment.py]: Also, this should be a private function (_rescaled_mean)

File: fmralign/template_alignment.py [hunk 53]
```diff
@@ -0,0 +1,349 @@
...
+           number of iterations in the alternate minimization. Each img is
+           aligned n_iter times to the evolving template. If n_iter = 0,
+           the template is simply the mean of the input images.
+        All other arguments are the same are passed to PairwiseAlignment
+    '''
+
+    aligned_imgs = imgs
+    template_history = []
+    iter = 0
+    while True:
```
Comment (bthirion) [inline fmralign/template_alignment.py]: why not a for loop ?

File: fmralign/template_alignment.py [hunk 54]
```diff
@@ -0,0 +1,349 @@
...
+                             self.alignment_method, self.n_pieces,
+                             self.clustering_method, self.n_bags,
+                             self.masker_, self.memory, self.memory_level,
+                             self.n_jobs, self.parallel_backend, self.verbose)
+        if save_template is not None:
+            self.template.to_filename(save_template)
+
+    def transform(self, imgs, train_index, test_index):
+        """
+        Predict data from X
```
Comment (bthirion) [inline fmralign/template_alignment.py]: What is X ?

File: fmralign/template_alignment.py (hunk 52 above)
Comment (thomasbazeille) [reply fmralign/template_alignment.py]: Your first comment is inexact, I recoded the function because I needed it :
mean_img docstring reads "Compute the mean of the images (in the time dimension of 4th dimension)" whereas euclidean_mean can handle lists of 4d images and compute their mean in the subject dimension which is the behavior I use by default.

Ok to rename it rescaled_mean, even though this name reflects a behavior that is not used most of the time.

Not sure it should be private. In every example / experiment with template, this will make the baseline mean of images across subjects

Comment (bthirion) [reply fmralign/template_alignment.py]: But why not call `mean_img` in those cases ?

Comment (thomasbazeille) [reply fmralign/template_alignment.py]: ... see my previous comment, they don't have the same behavior on lists of 4D Niimgs.

`mean_img` will return a 3D Niimg as nilearn docs state :
> Note that if list of 4D images are given, the mean of each 4D image is computed separately, and the resulting mean is computed after.

Unlike that, `euclidian_mean` will return a 4D Niimgs (just averaging across subject **i.e. the axis 0** from the list)  : 
```
def euclidean_mean(imgs, masker, scale_template=False):
    masked_imgs = [masker.transform(img) for img in imgs]
    average_img = np.mean(masked_imgs, axis=0)
```

Comment (thomasbazeille) [reply fmralign/template_alignment.py]: anyway, I took your advices

**Comment by thomasbazeille** [comment]
@bthirion all latest comments were adressed please review

File: fmralign/template_alignment.py [hunk 55]
```diff
@@ -0,0 +1,353 @@
...
+        self.low_pass = low_pass
+        self.high_pass = high_pass
+        self.t_r = t_r
+        self.memory = memory
+        self.memory_level = memory_level
+        self.n_jobs = n_jobs
+        self.parallel_backend = parallel_backend
+        self.verbose = verbose
+
+    def fit(self, imgs, scale_template=False, n_iter=2, save_template=None):
```
Comment (bthirion) [inline fmralign/template_alignment.py]: `scale_template`, `n_iter`, and `save_template` shall not be provided during fit but during class call.

File: fmralign/template_alignment.py [hunk 56]
```diff
@@ -0,0 +1,353 @@
...
+           source data. Every img must have the same length (number of sample)
+        scale_template: boolean
+            rescale template after each inference so that it keeps
+            the same norm as the average of training images.
+        n_iter: int
+           number of iterations in the alternate minimization. Each img is
+           aligned n_iter times to the evolving template. If n_iter = 0,
+           the template is simply the mean of the input images.
+        All other arguments are the same are passed to PairwiseAlignment
+    '''
```
Comment (bthirion) [inline fmralign/template_alignment.py]: Why not the return arguments ?

Comment (thomasbazeille) [reply fmralign/template_alignment.py]: I'm not sure I got your question right.

Comment (bthirion) [reply fmralign/template_alignment.py]: You describe the parameters not the return arguments.

**Comment by thomasbazeille** [comment]
@bthirion Last comments adressed, please review and merge if possible.

File: examples/template_alignment.py [hunk 57]
```diff
@@ -0,0 +1,179 @@
...
+#   towards the template
+# * target test: PA contrasts for subject sub-07, used as a ground truth to  \
+#   score our predictions
+#
+from nilearn.image import concat_imgs
+
+# We make a list of 4D niimgs from our list of list of files containing 3D images
+template_train = []
+for i in range(5):
+    template_train.append(concat_imgs(imgs[i]))
```
Comment (bthirion) [inline examples/template_alignment.py]:
```suggestion
    template_train = [concat_imgs(imgs[i]) for i in range(5)]
```

File: examples/template_alignment.py [hunk 58]
```diff
@@ -0,0 +1,179 @@
...
+#############################################################################
+# Defining a metric to score the performance of these predictions
+# ---------------------------------------------------
+# We define a scoring function to measure the correlation, between
+# the prediction and the ground truth for each voxel profiles of activation.
+
+import numpy as np
+from scipy.stats import pearsonr
+
+
```
Comment (bthirion) [inline examples/template_alignment.py]: I believe that such a metric cold be used elsewhere. Hence I suggest to integrate it to the library, e.g. in an `_utils.py` module.

Comment (bthirion) [reply examples/template_alignment.py]: And the, add a test ;-)

File: fmralign/template_alignment.py [hunk 59]
```diff
@@ -0,0 +1,389 @@
...
+        """
+        try:
+            template_length = self.template.shape[-1]
+            if not (all(i < template_length for i in test_index) and all(
+                    i < template_length for i in train_index)):
+                raise ValueError(
+                    ("Template has {} images but you provided a greater index in train_index or test_index.").format(template_length))
+        except (ValueError):
+            exit('Could not complete request.')
+        """
```
Comment (bthirion) [inline fmralign/template_alignment.py]: Remove that code

**Comment by thomasbazeille** [comment]
All changes mentionned are included, thanks for the review. To adress other issues on the list I would prefer to wait for this PR to be merged. 
These methods are used in my experiments for months now and seem robust, and concerning the example I think it is reasonably polished given it might change when improving documentation.
So I think it is time to merge.
